libs/remote_storage/src/s3_bucket.rs

```diff
@@ -37,82 +38,249 @@ use crate::{
 const MAX_DELETE_OBJECTS_REQUEST_SIZE: usize = 1000;
 
 pub(super) mod metrics {
-    use metrics::{register_int_counter_vec, IntCounterVec};
+    use metrics::{register_histogram_vec, register_int_counter_vec, Histogram, IntCounter};
     use once_cell::sync::Lazy;
 
-    static S3_REQUESTS_COUNT: Lazy<IntCounterVec> = Lazy::new(|| {
-        register_int_counter_vec!(
-            "remote_storage_s3_requests_count",
-            "Number of s3 requests of particular type",
-            &["request_type"],
-        )
-        .expect("failed to define a metric")
-    });
-
-    static S3_REQUESTS_FAIL_COUNT: Lazy<IntCounterVec> = Lazy::new(|| {
-        register_int_counter_vec!(
-            "remote_storage_s3_failures_count",
-            "Number of failed s3 requests of particular type",
-            &["request_type"],
-        )
-        .expect("failed to define a metric")
-    });
+    pub(super) static BUCKET_METRICS: Lazy<BucketMetrics> = Lazy::new(Default::default);
+
+    #[derive(Clone, Copy, Debug)]
+    pub(super) enum RequestKind {
+        Get = 0,
+        Put = 1,
+        Delete = 2,
+        List = 3,
+    }
+
+    use RequestKind::*;
+
+    impl RequestKind {
+        const fn as_str(&self) -> &'static str {
+            match self {
+                Get => "get_object",
+                Put => "put_object",
+                Delete => "delete_object",
+                List => "list_objects",
+            }
+        }
+        const fn as_index(&self) -> usize {
+            *self as usize
+        }
+    }
+
+    pub(super) struct RequestTyped<C>([C; 4]);
+
+    impl<C> RequestTyped<C> {
+        pub(super) fn get(&self, kind: RequestKind) -> &C {
+            &self.0[kind.as_index()]
+        }
+
+        fn build_with(mut f: impl FnMut(RequestKind) -> C) -> Self {
+            use RequestKind::*;
+            let mut it = [Get, Put, Delete, List].into_iter();
+            let arr = std::array::from_fn::<C, 4, _>(|index| {
+                let next = it.next().unwrap();
+                assert_eq!(index, next.as_index());
+                f(next)
+            });
+
+            if let Some(next) = it.next() {
+                panic!("unexpected {next:?}");
+            }
+
+            RequestTyped(arr)
+        }
+    }
+
+    impl RequestTyped<Histogram> {
+        pub(super) fn observe_elapsed(&self, kind: RequestKind, started_at: std::time::Instant) {
+            self.get(kind).observe(started_at.elapsed().as_secs_f64())
+        }
+    }
+
+    pub(super) struct PassFailCancelledRequestTyped<C> {
+        success: RequestTyped<C>,
+        fail: RequestTyped<C>,
+        cancelled: RequestTyped<C>,
```

Comment:
Do we need to have separate metric for cancelled requests? Do you have a use case for it in mind?

Reply:
I will add these drops in #4781, and we currently have (by my mistake) some FuturesUnordered downloading and I think we might have others as well.

Reply:
I mean this way we would know. If we have only zero in prod, then we can remove this.

Reply:
Works for me, maybe a comment would be nice, feel free to resolve